### PR TITLE
LinalgRefactor - SGVector&SGMatrix - setconst

### PR DESCRIPTION
**Pull request by OXPHOS**

### CODE REVIEW

File: tests/unit/mathematics/linalg/operations/Viennacl_operations_unittest.cc [hunk 1]
```diff
@@ -49,4 +49,23 @@ TEST(LinalgBackendViennaCL, dot)
...
+	for (index_t i = 0; i < size; ++i)
+		a[i] = value;
+
+	b_gpu = to_gpu(b);
+	set_const(b_gpu, value);
+	b_cpu = from_gpu(b_gpu);
+
+	for (index_t i = 0; i < size; ++i)
+		EXPECT_NEAR(a[i], b_cpu[i], 1E-15);
+}
```
Comment (OXPHOS) [inline tests/unit/mathematics/linalg/operations/Viennacl_operations_unittest.cc]: This doesn't look very right..
To set a gpu vector, ideally an empty gpu vector with malloced space should be provided.
However, there's no way to initiate such gpu vector from the SGVector interface.

Now, a cpu vector with unknown content is transferred to gpu and the gpu memory is reset by `value`.

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 2]
```diff
@@ -111,6 +133,14 @@ class LinalgBackendEigen : public LinalgBackendBase
 		return (typename SGVector<T>::EigenVectorXtMap(a)).dot(typename SGVector<T>::EigenVectorXtMap(b));
 	}
 
+	/** Set const method */
+	template <typename T>
+	void set_const_impl(SGVector<T>& vec, T value) const
+	{
+		for (index_t i = 0; i < vec.size(); ++i)
+			vec[i] = value;
```
Comment (lambday) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: isn't `std::fill` faster?

File: tests/unit/mathematics/linalg/operations/Viennacl_operations_unittest.cc (hunk 1 above)
Comment (lambday) [inline tests/unit/mathematics/linalg/operations/Viennacl_operations_unittest.cc]: do you get "reading uninitialized values" error on valgrind?
however, for this unit-test's purpose, you can set the value const to 0 for the cpu vector, transfer to gpu, set the value to something else, transfer back and test. that will get rid of this error.

File: src/shogun/mathematics/linalg/LinalgBackendBase.h [hunk 3]
```diff
@@ -104,6 +104,33 @@ class LinalgBackendBase
 	#undef BACKEND_GENERIC_DOT
 
 	/**
+	 * Wrapper method that set const values to vectors.
+	 *
+	 * @see linalg::set_const
+	 */
+	#define BACKEND_GENERIC_SET_CONST(Type) \
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendBase.h]: This will also be available for both vector and matrix I guess?

**Comment by karlnapf** [comment]
Generally fine, but pls also apply the changes discussed in the matrix PR here

**Comment by OXPHOS** [comment]
@vigsterkr ready to merge

**Pull request merged**